### PR TITLE
CSHARP-3257: Add .NET Standard 2.1 support.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/5fbd5b220305b931953b41db

### CODE REVIEW

File: src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 using System;
 using System.Collections.Generic;
-#if NET452 || NETSTANDARD2_0
+#if !NETSTANDARD1_5
```
Comment (DmitryLukyanov) [inline src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs:18]: make this change to pass tests

File: tests/MongoDB.Bson.Tests/Jira/CSharp301Tests.cs [hunk 2]
```diff
@@ -60,7 +60,10 @@ public void TestHashtableEmpty()
         {
             var c = new C { Id = 1, Obj = new Hashtable { } };
             var json = c.ToJson();
-#if NET452
+#if NET452 || NETCOREAPP3_0
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Bson.Tests/Jira/CSharp301Tests.cs:63]: the whole logic about well-known libraries looks problematic. For example in coreapp3.0(not only here) the list of well-known assemblies looks like:

    [0]: {System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e}
    [1]: {System.Collections, Version=4.0.12.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [2]: {System.Collections, Version=4.0.12.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}

File: src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs (hunk 1 above)
Comment (rstam) [inline src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs:18]: I was expecting 

    #if NET452 || NETSTANDARD2_0 || NETSTANDARD2_1

Negative `#if` kind of sucks in general, and they are also dangerous (in essence rather than listing which frameworks this feature is supported in your are stating that it will be supported in all frameworks present and future other than 1.5)

But OK...

File: tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs [hunk 3]
```diff
@@ -1117,6 +1117,7 @@ public class TestClass
             public float S;
         }
 
+#if !NETCOREAPP3_0
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs]: this is related to this ticket: https://jira.mongodb.org/browse/CSHARP-3267

File: tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs [hunk 4]
```diff
@@ -91,7 +91,8 @@ public void Execute_should_use_cached_IWireProtocol_if_available([Values(false,
             else
             {
                 var e = exception.Should().BeOfType<ArgumentException>().Subject;
-                e.Message.Should().Be("CommandResponseHandling must be Return, NoneExpected or ExhaustAllowed.\r\nParameter name: responseHandling");
+                e.Message.Should().StartWith("CommandResponseHandling must be Return, NoneExpected or ExhaustAllowed.");
+                e.ParamName.Should().Be("responseHandling");
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs:95]: How is this related to CSHARP-3257?

File: tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs [hunk 5]
```diff
@@ -838,8 +838,8 @@ public void TestRename()
             Assert.Equal(expected, update.ToJson());
         }
 
-        // tests fail on .NET Core 2.1 with out of order key-value-pairs        
-#if !NETCOREAPP2_1
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs:842]: An example of why negative `#if` are dangerous!

File: tests/MongoDB.Driver.Tests/Linq/Translators/AggregateProjectTranslatorTests.cs [hunk 6]
```diff
@@ -1347,7 +1347,7 @@ public void Should_translate_split()
         {
             RequireServer.Check().VersionGreaterThanOrEqualTo("3.3.6");
 
-#if !NETCOREAPP2_1
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Translators/AggregateProjectTranslatorTests.cs:1350]: Another example of why negative #if are dangerous!

File: tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs [hunk 7]
```diff
@@ -91,7 +91,8 @@ public void Execute_should_use_cached_IWireProtocol_if_available([Values(false,
             else
             {
                 var e = exception.Should().BeOfType<ArgumentException>().Subject;
-                e.Message.Should().Be("CommandResponseHandling must be Return, NoneExpected or ExhaustAllowed.\r\nParameter name: responseHandling");
+                e.Message.Should().StartWith("CommandResponseHandling must be Return, NoneExpected or ExhaustAllowed.");
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs:94]: .NET Core 3 uses a bit different format in the final message regarding including `ParamName` information. So splitting it into two validations solved this issue

File: tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs [hunk 8]
```diff
@@ -838,8 +838,8 @@ public void TestRename()
...
+        // tests fail on .NET Core 2.1 and higher frameworks with out of order key-value-pairs
+#if NET452 && NETCOREAPP1_1
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs]: It looks like all next frameworks will have behavior which is currently problematic for this test

File: src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs:18]: I think the main idea is about excluding the old framework. So, `!` means exactly `exclude the framework` with different behavior

File: tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs:95]: otherwise this test fails

File: tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs:842]: I think the idea of this test (at least now) is to check behavior which supported only on older frameworks, so new frameworks are not involved in this test

File: src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs:18]: I think there is no reason to assume that this functionality will be excluded in the future, but even if so, it will be a compile error

File: tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs (hunk 4 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs:95]: OK. Thanks for the explanation.

File: tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs (hunk 8 above)
Comment (rstam) [reply tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs]: This `#if` can't be right. 

Only one of the two target frameworks can be defined at the same time.

File: tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs (hunk 5 above)
Comment (rstam) [reply tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs:842]: I'm just pointing out that negative tests are dangerous because you don't know what the future holds.

File: src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs:18]: Agreed. But negative `#if` is dangerous.

File: tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs]: agreed. My fault

Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs]: done

**Review by durran — APPROVED** [review]

**Review by rstam — COMMENTED** [review]
I have no further changes to request, but I am going to wait and see what comments other reviewers have before LGTM'ing.

File: src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs (hunk 1 above)
Comment (JamesKovacs) [reply src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs:18]: While I'm not a big fan of negative `#if` statements either, it makes sense to exclude the old framework with different behaviour. Then when we retire support for `NETSTANDARD1_5`, we can simply find the usages of `NETSTANDARD1_5` and remove them from the code.

File: tests/MongoDB.Bson.Tests/Jira/CSharp301Tests.cs [hunk 9]
```diff
@@ -60,7 +60,10 @@ public void TestHashtableEmpty()
...
+            // Hashtable is situated in well-known libraries for:
+            // - NET452: mscorlib
+            // - NETCOREAPP3_0: System.Private.CoreLib
             var discriminator = "System.Collections.Hashtable";
 #else
             var discriminator = typeof(Hashtable).AssemblyQualifiedName;
```
Comment (JamesKovacs) [inline tests/MongoDB.Bson.Tests/Jira/CSharp301Tests.cs:69]: This is an ugly problem, but unrelated to this PR. We shouldn't be using a default discriminator that changes from one framework version to another. But I suppose that ship already sailed long ago.

Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/Jira/CSharp301Tests.cs:69]: agreed, I mentioned it here: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/162#discussion_r529840987

Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/Jira/CSharp301Tests.cs:69]: not sure whether we want to make any changes here (in the scope of a different ticket). Most-likely no

File: tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs (hunk 3 above)
Comment (JamesKovacs) [reply tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs]: Rather than commenting out the entire test for NETCOREAPP3_0, I'd rather see us provide the correct expected result along with a comment referencing CSHARP-3267:
```
#if NETCOREAPP3_0
            expected = expected.Replace("#S", "-3.4028235E+38");
#else
            expected = expected.Replace("#S", "-3.40282347E+38");
#endif
```

File: tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs [hunk 10]
```diff
@@ -1215,6 +1217,7 @@ public void TestOnePointFive()
             Assert.True(bson.SequenceEqual(rehydrated.ToBson()));
         }
 
+#if !NETCOREAPP3_0
```
Comment (JamesKovacs) [inline tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs]: Same comment as for `TestMin`. Let's change our expected result rather than omitting the test entirely under NETCOREAPP3_0.

File: tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs (hunk 5 above)
Comment (JamesKovacs) [reply tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs:842]: If the test is assuming ordered key/value pairs - which a hashtable or dictionary won't necessarily have - then we should fix the test rather than commenting it out for some targets.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Overall looks good. I don't like that we are commenting out tests for certain targets. If the test is broken for certain frameworks, that is an indication of a fragile test and we should fix it.

File: tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs:842]: this was commented a long time, I think we have a ticket to fix it (which was discussed several times on grooming, and then we moved it into backlog as `Open`).

Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs:842]: so, I don't think that we need to work on it now

File: tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs]: done

File: tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs (hunk 10 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/Serialization/Serializers/NetPrimitiveSerializerTests.cs]: done

**Comment by DmitryLukyanov** [comment]
ready for review

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Review by rstam — APPROVED** [review]
LGTM
